packages/@aws-cdk/aws-stepfunctions-tasks/lib/dynamodb/shared-types.ts

```diff
@@ -0,0 +1,191 @@
+ */
+export enum DynamoItemCollectionMetrics {
+  /**
+   * If set to SIZE, the response includes statistics about item collections,
+   * if any, that were modified during the operation.
+   */
+  SIZE = 'SIZE',
+
+  /**
+   * If set to NONE, no statistics are returned.
+   */
+  NONE = 'NONE',
+}
+
+/**
+ * Use ReturnValues if you want to get the item attributes as they appear before or after they are changed
+ */
+export enum DynamoReturnValues {
+  /**
+   * Nothing is returned
+   */
+  NONE = 'NONE',
+
+  /**
+   * Returns all of the attributes of the item
+   */
+  ALL_OLD = 'ALL_OLD',
+
+  /**
+   * Returns only the updated attributes
+   */
+  UPDATED_OLD = 'UPDATED_OLD',
+
+  /**
+   * Returns all of the attributes of the item
+   */
+  ALL_NEW = 'ALL_NEW',
+
+  /**
+   * Returns only the updated attributes
+   */
+  UPDATED_NEW = 'UPDATED_NEW',
+}
+
+/**
+ * Property for any key
+ */
+export interface DynamoAttribute {
+  /**
+   * The name of the attribute
+   */
+  readonly name: string;
+
+  /**
+   * The value of the attribute
+   */
+  readonly value: DynamoAttributeValue;
+}
+
+/**
+ * Class to generate projection expression
+ */
+export class DynamoProjectionExpression {
+  private expression: string[] = [];
+
+  /**
+   * Adds the passed attribute to the chain
+   *
+   * @param attr Attribute name
+   */
+  public withAttribute(attr: string): DynamoProjectionExpression {
+    if (this.expression.length) {
+      this.expression.push(`.${attr}`);
+    } else {
+      this.expression.push(attr);
+    }
+    return this;
+  }
+
+  /**
+   * Adds the array literal access for passed index
+   *
+   * @param index array index
+   */
+  public atIndex(index: number): DynamoProjectionExpression {
+    if (!this.expression.length) {
+      throw new Error('Expression must start with an attribute');
+    }
+
+    this.expression.push(`[${index}]`);
+    return this;
+  }
+
+  /**
+   * converts and return the string expression
+   */
+  public toString(): string {
+    return this.expression.join('');
+  }
+}
+
+/**
+ * Map of string to AttributeValue
+ */
+export type DynamoAttributeValueMap = {
+  [key: string]: DynamoAttributeValue;
+};
+
+/**
+ * Class to generate AttributeValue
+ */
+export interface DynamoAttributeValue {
+  /**
+   * Sets an attribute of type String. For example:  "S": "Hello"
+   */
+  readonly s?: string;
+
+  /**
+   * Sets an attribute of type Number. For example:  "N": "123.45"
+   * Numbers are sent across the network to DynamoDB as strings,
+   * to maximize compatibility across languages and libraries.
+   * However, DynamoDB treats them as number type attributes for mathematical operations.
+   */
+  readonly n?: string;
+
+  /**
+   * Sets an attribute of type Binary. For example:  "B": "dGhpcyB0ZXh0IGlzIGJhc2U2NC1lbmNvZGVk"
+   */
+  readonly b?: string;
+
+  /**
+   * Sets an attribute of type String Set. For example:  "SS": ["Giraffe", "Hippo" ,"Zebra"]
+   */
+  readonly ss?: string[];
```

Comment:
need help with a friendlier naming convention here. are these too ambiguous? 

should we go with something like `stringValue`, `numberValue` ...
open to any suggestions

Reply:
These acronyms are pretty common in DDB world - [JS SDK](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB.html#getItem-property) and [CLI](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/getting-started-step-2.html). Might be worth keeping them as is.

Reply:
How about something closer to the GetItem API on the SDK - https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB.html#getItem-property

```ts
new tasks.DynamoGetItem(this, 'Get Item', {
  key: {
    messageId: {
      s: 'message-007' 
    }
  },
  tableName: 'my-table',
});
```

Reply:
Personally find this API gross and ill-suited for typed contexts (it does make sense in the mostly untyped world of JSON).

To a large extent I'd much rather have `ddb.AttributeValue.fromString("message-007")`. In particular, the `messageId` in this case could only have ONE of the types (`s`,  `n`, `b`, ...), which would imply a fully type-safe form would require a type union. And that's going to get you in trouble sooner or later.

Reply:
@RomainMuller fair point - i'll take a run at your suggestion!